code/modules/projectiles/gun.dm

```diff
@@ -395,38 +388,52 @@
 	else
 		return ..()
 
+/obj/item/gun/proc/being_worked_on()
+	return (!isturf(src.loc) || !(locate(/obj/structure/rack) in src.loc))
+
+/obj/item/gun/proc/work_on()		
+	var/list/possible_items = list()
+
+	if(can_flashlight && gun_light)
+		LAZYADD(possible_items,gun_light)
+	if(can_bayonet && bayonet)
+		LAZYADD(possible_items,bayonet)
+	if(pin)
+		LAZYADD(possible_items,pin)
+	
+	var/obj/item/item_to_remove = input(user, "Select an attachment to remove", "Attachment Removal") as null|obj in possible_items
+	if(!item_to_remove || !user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
+		return
+	if(I.use_tool(src, user, FIRING_PIN_REMOVAL_DELAY, volume = 50))
+		return remove_gun_attachment(user, I, item_to_remove)
+
 /obj/item/gun/screwdriver_act(mob/living/user, obj/item/I)
 	. = ..()
 	if(.)
 		return
 	if(!user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
 		return
-	if((can_flashlight && gun_light) && (can_bayonet && bayonet)) //give them a choice instead of removing both
-		var/list/possible_items = list(gun_light, bayonet)
-		var/obj/item/item_to_remove = input(user, "Select an attachment to remove", "Attachment Removal") as null|obj in possible_items
-		if(!item_to_remove || !user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
-			return
-		return remove_gun_attachment(user, I, item_to_remove)
-
-	else if(gun_light && can_flashlight) //if it has a gun_light and can_flashlight is false, the flashlight is permanently attached.
-		return remove_gun_attachment(user, I, gun_light, "unscrewed")
-
-	else if(bayonet && can_bayonet) //if it has a bayonet, and the bayonet can be removed
-		return remove_gun_attachment(user, I, bayonet, "unfix")
-
+	if ( being_worked_on() )	
+		to_chat(user, "<span class='warning'>You only modify the [src] on a weapon crafting table!</span>")
+		return FALSE
+	work_on()
+		
 /obj/item/gun/proc/remove_gun_attachment(mob/living/user, obj/item/tool_item, obj/item/item_to_remove, removal_verb)
 	if(tool_item)
 		tool_item.play_tool_sound(src)
 	to_chat(user, "<span class='notice'>You [removal_verb ? removal_verb : "remove"] [item_to_remove] from [src].</span>")
 	item_to_remove.forceMove(drop_location())
 
-	if(Adjacent(user) && !issilicon(user))
-		user.put_in_hands(item_to_remove)
+	if (Adjacent(user) && !issilicon(user))
+		put_in_hands(item_to_remove)
 
 	if(item_to_remove == bayonet)
 		return clear_bayonet()
 	else if(item_to_remove == gun_light)
 		return clear_gunlight()
+	else if(item_to_remove == pin)
+		QDEL_NULL(pin)
+		return TRUE
```

Comment:
You should probably use a switch statement here

Reply:
tried; switch doesn't work cause not constant.

Reply:
oh, right it's a var, I'm duuumb. my bad